src/receive/mod.rs

```diff
@@ -0,0 +1,434 @@
+            }
+        }) {
+            Ok(e) => Ok(Arc::new(e.into())),
+            Err(e) => Err(e.into()),
+        }
+    }
+}
+
+///A mutable checked proposal that the receiver may contribute inputs to make a payjoin.
+pub struct ProvisionalProposal(Mutex<PdkProvisionalProposal>);
+
+impl From<PdkProvisionalProposal> for ProvisionalProposal {
+    fn from(value: PdkProvisionalProposal) -> Self {
+        Self(Mutex::new(value))
+    }
+}
+
+pub trait ProcessPartiallySignedTransaction {
+    fn callback(&self, psbt: String) -> Result<String, PayjoinError>;
+}
+
+impl ProvisionalProposal {
+    fn mutex_guard(&self) -> MutexGuard<'_, payjoin::receive::ProvisionalProposal> {
+        self.0.lock().unwrap()
+    }
+    pub fn substitute_output_address(
+        &self,
+        substitute_address: String,
+    ) -> Result<(), PayjoinError> {
+        let address =
+            payjoin::bitcoin::Address::from_str(substitute_address.as_str())?.assume_checked();
+        Ok(self.mutex_guard().substitute_output_address(address))
+    }
+    pub fn contribute_witness_input(
+        &self,
+        txo: TxOut,
+        outpoint: OutPoint,
+    ) -> Result<(), PayjoinError> {
+        let txo: payjoin::bitcoin::blockdata::transaction::TxOut = txo.into();
+        Ok(self.mutex_guard().contribute_witness_input(txo, outpoint.into()))
+    }
+
+    pub fn contribute_non_witness_input(
+        &self,
+        tx: Vec<u8>,
+        outpoint: OutPoint,
+    ) -> Result<(), PayjoinError> {
+        let tx: payjoin::bitcoin::Transaction =
+            payjoin::bitcoin::consensus::encode::deserialize(&*tx)?;
+        Ok(self.mutex_guard().contribute_non_witness_input(tx, outpoint.into()))
+    }
+
+    /// Select receiver input such that the payjoin avoids surveillance. Return the input chosen that has been applied to the Proposal.
+    ///
+    /// Proper coin selection allows payjoin to resemble ordinary transactions. To ensure the resemblance, a number of heuristics must be avoided.
+    ///
+    /// UIH “Unnecessary input heuristic” is one class of them to avoid. We define UIH1 and UIH2 according to the BlockSci practice BlockSci UIH1 and UIH2:
+    pub fn try_preserving_privacy(
+        &self,
+        candidate_inputs: HashMap<u64, OutPoint>,
+    ) -> Result<OutPoint, PayjoinError> {
+        let mut _candidate_inputs: HashMap<payjoin::bitcoin::Amount, payjoin::bitcoin::OutPoint> =
+            HashMap::new();
+        for (key, value) in candidate_inputs.iter() {
+            _candidate_inputs.insert(
+                payjoin::bitcoin::Amount::from_sat(key.to_owned()),
+                value.to_owned().into(),
+            );
+        }
+
+        match self.mutex_guard().try_preserving_privacy(_candidate_inputs) {
+            Ok(e) => Ok(OutPoint { txid: e.txid.to_string(), vout: e.vout }),
+            Err(e) => Err(PayjoinError::SelectionError { message: format!("{:?}", e) }),
+        }
+    }
```

Comment:
My understanding is underscore prefix is usually for unused variables. Variable shadowing could work here, as could `.map()`.

```suggestion
   pub fn try_preserving_privacy(
        &self,
        candidate_inputs: HashMap<u64, OutPoint>,
    ) -> Result<OutPoint, PayjoinError> {
        use payjoin::bitcoin::Amount;

        // Transform `candidate_inputs` directly into the desired HashMap
        let candidate_inputs: HashMap<Amount, payjoin::bitcoin::OutPoint> = candidate_inputs.into_iter()
            .map(|(key, value)| (Amount::from_sat(key), value.into()))
            .collect();

        match self.mutex_guard().try_preserving_privacy(candidate_inputs) {
            Ok(e) => Ok(OutPoint { txid: e.txid.to_string(), vout: e.vout }),
            Err(e) => Err(PayjoinError::SelectionError { message: format!("{:?}", e) }),
        }
    }
```

Reply:
@DanGould Thanks, suggestion has been implemented